# The question before the budget

**Week 1, Monday. Notebook 1 of 4.** By the end of this notebook you can take a CEO's question,
say what "sales" means before you total it, and draw revenue as a tree whose every branch is a
number someone can measure and move.

> **The client asks.** "Before I sign anything, I want to understand our own sales. What is 'sales'
> made of? Where does revenue come from, by customer type and channel? Is acquisition even the
> branch that is short?"
>
> Meera Raghavan, CEO, Kalpa Retail

You are the newest engineers in the data and AI team at Kalpa's Global Capability Centre in
Bengaluru, the in-house centre that builds data and AI systems for all five of Kalpa Group's
business units. Kalpa Retail sells consumer goods through its app, its website and its stores
across India and South-East Asia. Revenue grew 4 percent last year against a plan of 15, the board
wants a growth plan within a month, and marketing has asked for Rs 12 crore to acquire new
customers. Meera's question reached your team before anyone opened a tool, and this notebook is
the thinking that comes before the tool.

**Setup.** The next cell finds `scripts/c2kit.py`, the programme's shared helper, by walking up
from this notebook's folder, so it works wherever the repository is cloned. Then it loads the day's
thirty orders from `../data/`, the only place the data lives. Run it first: every later cell uses
the names it makes.

In [1]:
import sys, pathlib
here = pathlib.Path.cwd()
for parent in [here, *here.parents]:
    if (parent / "scripts" / "c2kit.py").exists():
        sys.path.insert(0, str(parent / "scripts")); break
import c2kit as kit

ORDERS = kit.load_records("C2_W01_D01_orders_STUDENT.py")
print(len(ORDERS), "Kalpa Retail orders loaded, placed between 1 July and 26 September 2026")

30 Kalpa Retail orders loaded, placed between 1 July and 26 September 2026


In [2]:
# The map: where this notebook sits in the day, and what it adds, in order.
kit.side_by_side(
    kit.ladder(["The question before the budget", "The first count", "The leaves", "The average that lies", "Hands-on: your call"], lit=0, show=False),
    kit.vflow(["the ask\nfour questions in one message",
               "the readings of sales\nfour honest totals",
               "the tree\nrevenue as multiplied parts",
               "the bills\nwhat each branch costs to move",
               "the bet\nwhere the Rs 12 crore lands"], show=False),
)

## 1. The ask, before any tool

A client message usually carries several questions, and an engineer's first job is to separate
them and say when each will be answered. Meera's message carries three. Anand Iyer, the finance
controller, adds a fourth that is really a warning about the data: "No averages. One business
customer can move an average."

| What she asked | What it becomes for the team | Answered |
|---|---|---|
| What is sales made of | The revenue tree, with every branch counted | Today |
| By customer type and channel | Revenue grouped by segment and by channel | Tuesday |
| Is acquisition the short branch | Which branch moved between two quarters | Tuesday |
| No averages | The typical order, which one record cannot move | Today, notebook 4 |

Two of the four belong to today. An analyst who tries to answer all four on day one answers none
of them well, so the table is the first deliverable: Meera sees what she will get, and when.

The same five moves turn every business ask into work, and they run in this order before a tool
opens.

In [3]:
kit.flow(["the ask\nbusiness words",
          "the question\nwhich parts multiply",
          "the data\none row per order",
          "the method\nbreak it, count each part",
          "the answer\na number per branch"],
         lit=1, title="From a business ask to an answer: the lit move is the one this day trains")

### Predict before you run: which total is "sales"?

Meera will read "sales" in the note the team sends on Thursday. The next cell totals the same
thirty orders three ways. Before you run it, pick the total that belongs in her note:

a) the biggest; b) the most conservative; c) any one of them, with its definition written beside
it; d) none, until the discounts are known.

The cell uses a loop and an `if`, which notebooks 2 and 3 teach. Read its output now and come back
to its code tonight; that second reading is the point.

In [4]:
# Three totals from the same thirty orders. Why int() is needed here is notebook 2's lesson.
totals = {"booked": 0, "not cancelled": 0, "delivered": 0}
counts = {"booked": 0, "not cancelled": 0, "delivered": 0}
for order in ORDERS:
    amount = int(order["amount"])
    totals["booked"] += amount
    counts["booked"] += 1
    if order["status"] != "cancelled":
        totals["not cancelled"] += amount
        counts["not cancelled"] += 1
    if order["status"] == "delivered":
        totals["delivered"] += amount
        counts["delivered"] += 1

kit.vflow([f"booked\n{counts['booked']} orders, {kit.rupees(totals['booked'])}",
           f"not cancelled\n{counts['not cancelled']} orders, {kit.rupees(totals['not cancelled'])}",
           f"delivered\n{counts['delivered']} orders, {kit.rupees(totals['delivered'])}",
           "after discounts\nno discount field, so no number yet"],
          kinds=[None, None, None, "unknown"],
          edges=["less the cancelled", "less the returned", "less what was given back"],
          title="One word, four readings: every arrow removes something")

**What happened.** All four readings are correct arithmetic, and each answers a different
question: how much demand reached Kalpa, how much it set out to fulfil, how much reached a customer
and stayed there, and how much it kept after discounts, which this file cannot say. The answer is
c. Option d sounds rigorous and is a way of never answering, because a number can be stated now
with a line saying what it leaves out.

> **Kavya's review.** Name the definition before the number, every time: "Revenue, all booked
> orders, 1 July to 26 September: Rs 5,44,810." Today's counts use every booked order, as the file
> arrives, and the note says so.

Kavya Nair is the senior analyst on the team. Her review is what a senior checks before any work
leaves the team, and it closes every section this week.

In [5]:
kit.check("each reading removes something, so the totals shrink in order",
          totals["booked"] >= totals["not cancelled"] >= totals["delivered"],
          " then ".join(kit.rupees(v) for v in totals.values()))
kit.check("all thirty orders are booked and twenty-one were delivered",
          counts["booked"] == 30 and counts["delivered"] == 21, str(counts))

## 2. The tree, drawn before any code

Revenue broken once is how many customers bought, times what each one brought. Break "what each
one brought" again, and again, and every leaf on the right of the tree is a count or a price that
somebody in the business can move. Discounts come off the top, so the multiplied part is gross
revenue.

This is the picture from today's slides and from the cheat sheet, and it is the most reused
drawing of the week. Read it left to right: each box is made of the boxes it points to.

In [6]:
TREE = {"label": "revenue", "note": "gross less discounts", "children": [
    {"label": "gross revenue", "note": "customers times spend", "children": [
        {"label": "customers", "note": "how many bought"},
        {"label": "revenue per customer", "note": "orders times order value", "children": [
            {"label": "orders per customer", "note": "how often each came back"},
            {"label": "revenue per order", "note": "items times price", "children": [
                {"label": "items per order", "note": "how full the basket was"},
                {"label": "price per item", "note": "what each line cost"}]}]}]},
    {"label": "discounts", "note": "what was given back"}]}

kit.driver_tree(TREE, title="The revenue tree: five leaves, and every one is a lever")

**Rigor: the units have to cancel.** A tree of multiplications is right only if the units work.
Customers times orders per customer gives orders; orders times items per order gives items; items
times price per item gives rupees. If a proposed branch does not fit that chain, it belongs
somewhere else on the tree, which is a quick way to catch a muddled metric before it reaches a
client.

In [7]:
kit.equation(["revenue", "=", "customers", "×", "orders per customer", "×", "items per order",
              "×", "price per item", "−", "discounts"], title="The same tree as one line")
kit.table(["Multiply", "by", "and you get"],
          [("customers", "orders per customer", "orders"),
           ("orders", "items per order", "items"),
           ("items", "price per item", "rupees before discounts"),
           ("rupees before discounts", "less discounts", "revenue")],
          caption="The units check: each line's result is the next line's input")

Multiply,by,and you get
customers,orders per customer,orders
orders,items per order,items
items,price per item,rupees before discounts
rupees before discounts,less discounts,revenue


### Every branch is a metric with a denominator

A rate without its denominator cannot be checked, compared or defended, so every branch is written
as a numerator over a denominator before anything is computed.

| Branch | Numerator | Denominator | From today's file |
|---|---|---|---|
| Customers | Distinct customer ids | None, it is a count | Yes |
| Orders per customer | Orders | Distinct customers | Yes |
| Items per order | Items sold | Orders | No items in this file |
| Price per item | Revenue before discounts | Items sold | No items in this file |
| Discounts | Rupees given back | Revenue before discounts | No discount field |

The three rows marked "no" are the most useful rows in the table. An engineer who can say what the
data cannot answer, and which fields would answer it, is more useful than one who fills the gap
with a guess.

> **Kavya's review.** Say the denominator out loud. Orders per customer is orders over customers,
> never the other way up, and a share of discounts is over revenue before discounts, never after.

### Five branches, five different bills

A 10 percent lift on any one branch lifts revenue by about 10 percent. What differs is what the
lift costs, and what can go wrong.

| Branch | How it is moved | What it costs | The risk |
|---|---|---|---|
| Customers | Acquisition | Marketing spend | New customers who buy once and never return |
| Orders per customer | Retention | Loyalty and service | Rewarding people who would have come back anyway |
| Items per order | Basket | Merchandising and bundles | Baskets filled with low-margin lines |
| Price per item | Price | Volume at risk | The price-sensitive leave first |
| Discounts | Margin | Margin traded for quantity | Volume that grows by less than the cut |

### Predict before you run: two 10 percent lifts

Orders per customer rises 10 percent and items per order rises 10 percent. Does revenue rise by 20
percent, by 21, or by something else? And a 15 percent discount that lifts the items sold by 10
percent: does revenue rise or fall?

In [8]:
both = 1.10 * 1.10         # two branches that multiply, each lifted 10 percent
discount = 1.10 * 0.85     # 10 percent more volume at 15 percent off
print(both)
print(round(both, 2))
print(round(discount, 3))

1.2100000000000002
1.21
0.935


**What happened.** Two 10 percent lifts give 21 percent, where adding them says 20, because the
branches multiply. The same multiplication works against you: 15 percent off that buys only 10
percent more volume leaves revenue at 0.935 of what it was, a fall of 6.5 percent.

The first line printed `1.2100000000000002`. That trailing 2 is how a binary floating point number
stores 1.1 times 1.1, and it is why a float is rounded for display and never compared with `==`.

In [9]:
kit.check("two 10 percent lifts compound to 21 percent", round(both, 2) == 1.21, f"{both:.4f}")
kit.check("the discount leaves revenue 6.5 percent lower", round(1 - discount, 3) == 0.065,
          f"{discount:.4f}")
kit.check("a float is compared after rounding, never with ==", both != 1.21 and round(both, 2) == 1.21,
          repr(both))

## 3. Where the Rs 12 crore lands

Marketing's budget is a bet on one leaf out of five, placed before anyone checked which leaf is
short. Of the gap between 4 and 15 percent, how much came from fewer customers, and how much from
each customer buying less? Tuesday's file answers that; today's job is to see that the question
exists.

In [10]:
BET = {"label": "revenue", "note": "4% against a plan of 15%", "children": [
    {"label": "gross revenue", "note": "customers times spend", "children": [
        {"label": "customers", "note": "Rs 12 crore bets here", "kind": "lit"},
        {"label": "revenue per customer", "note": "orders times order value", "children": [
            {"label": "orders per customer", "note": "how often each came back"},
            {"label": "revenue per order", "note": "items times price", "children": [
                {"label": "items per order", "note": "how full the basket was"},
                {"label": "price per item", "note": "what each line cost"}]}]}]},
    {"label": "discounts", "note": "what was given back"}]}

kit.driver_tree(BET, title="Marketing's Rs 12 crore is a bet on the first leaf")

### The placement drill, and its answer

Try it before reading the table: place each initiative on the branch it moves, and say in one line
what moving that branch costs.

| Initiative | The branch it moves | What moving it costs | The catch |
|---|---|---|---|
| A 15 percent discount, on everything for a month | Discounts, hoping for more items and orders | Margin | Volume must rise by more than the cut, or revenue falls |
| A new store, in a city where the app already sells | Customers | Rent and fit-out | It may move app customers into the store and add nobody |
| A loyalty card, points on every order | Orders per customer | Points and their cost | It pays people who would have come back anyway |
| A price rise, five percent on the top sellers | Price per item | Volume at risk | The price-sensitive leave first |
| An app redesign, a new checkout and home screen | Depends on the behaviour it changes | The build | Name the behaviour before placing it |

The redesign is the row that teaches. An initiative is placed by the behaviour it changes, never by
what it is called: a checkout with one step fewer lifts conversion, which is customers, and a home
screen with reorder buttons lifts frequency.

### In the interview

**[S] How would you increase sales for an online retailer?** A strong answer starts from the tree
rather than from tactics. "I would draw revenue as customers, times orders per customer, times
items per order, times price per item, less discounts, and find out which branch is short before
choosing how to move it. Each branch has a different bill: acquisition costs marketing, frequency
costs retention, basket costs merchandising, price risks volume, and discounts trade margin for
quantity. So I would compare each branch across two periods, find the one that moved, and propose
the cheapest lever on it, with the measure that will show whether it worked."

**[D] Marketing wants budget for acquisition; what would you check before agreeing it is the right
branch, and how would you say no?** "I would agree with the goal and then check whether customers
is the branch that is short: customers and orders per customer, this quarter against the last. If
customers held and frequency fell, acquisition money is aimed at a branch that is fine. The no
comes as a proposal: run that comparison first, which takes days and costs almost nothing, then
fund whichever branch it points at, with a measure agreed in advance."

In [11]:
kit.table(["What this notebook established", "The evidence"],
          [("A client message is several questions, each with its own answer date",
            "Meera's four questions, two of them today's"),
           ("A total is stated with its definition", "booked, not cancelled, delivered: three honest totals"),
           ("Revenue is a tree of multiplied parts, less discounts",
            "five leaves, and the units cancel down to rupees"),
           ("Every branch is a numerator over a denominator", "three of five leaves are not in this file"),
           ("Branches multiply, so small lifts compound", "1.10 times 1.10 is 1.21, and a weak discount loses 6.5 percent")],
          caption="Summary: the thinking that comes before the code")

What this notebook established,The evidence
"A client message is several questions, each with its own answer date","Meera's four questions, two of them today's"
A total is stated with its definition,"booked, not cancelled, delivered: three honest totals"
"Revenue is a tree of multiplied parts, less discounts","five leaves, and the units cancel down to rupees"
Every branch is a numerator over a denominator,three of five leaves are not in this file
"Branches multiply, so small lifts compound","1.10 times 1.10 is 1.21, and a weak discount loses 6.5 percent"


In [12]:
kit.check_summary()
print("Next: notebook 2 opens the workbench and counts the first leaf, until the count breaks.")

Next: notebook 2 opens the workbench and counts the first leaf, until the count breaks.
